river/cluster/denstream.py

```diff
@@ -43,411 +44,424 @@ class DenStream(base.Clusterer):
 
     >>> denstream = cluster.DenStream(decaying_factor = 0.01,
-    ...                               core_weight_threshold = 1.01,
-    ...                               tolerance_factor = 1.0005,
-    ...                               radius = 0.5)
+    ...                               beta = 1.01,
+    ...                               mu = 1.0005,
+    ...                               epsilon = 0.5,
+    ...                               n_samples_init=10)
 
     >>> for x, _ in stream.iter_array(X):
     ...     denstream = denstream.learn_one(x)
 
     >>> denstream.predict_one({0: -1, 1: -2})
-    0
+    1
 
     >>> denstream.predict_one({0:5, 1:4})
-    1
+    0
 
-    >>> denstream.n_clusters
+    >>> denstream.predict_one({0:1, 1:1})
     2
 
+    >>> denstream.n_clusters
+    3
+
     """
 
+    class BufferItem:
+        def __init__(self, x, timestamp, covered):
+            self.x = x
+            self.timestamp = (timestamp,)
+            self.covered = covered
+
     def __init__(
         self,
         decaying_factor: float = 0.25,
-        core_weight_threshold: float = 5,
-        tolerance_factor: float = 0.5,
-        radius: float = 2,
+        beta: float = 5,
+        mu: float = 0.5,
+        epsilon: float = 0.02,
+        n_samples_init: int = 1000,
+        stream_speed: int = 100,
     ):
         super().__init__()
-        self.time_stamp = -1
+        self.timestamp = -1
         self.initialized = False
         self.decaying_factor = decaying_factor
-        self.core_weight_threshold = core_weight_threshold
-        self.tolerance_factor = tolerance_factor
-        self.radius = radius
+        self.beta = beta
+        self.mu = mu
+        self.epsilon = epsilon
+        self.n_samples_init = n_samples_init
+        self.stream_speed = stream_speed
 
         # number of clusters generated by applying the variant of DBSCAN algorithm
-        # on p micro-cluster centers and their centers
+        # on p-micro-cluster centers and their centers
         self.n_clusters = 0
         self.clusters = {}
-        self.centers = {}
         self.p_micro_clusters = {}
         self.o_micro_clusters = {}
 
-    @property
-    def _time_point(self):
-        return math.ceil(
-            1
-            / self.decaying_factor
-            * math.log(
-                self.tolerance_factor
-                * self.core_weight_threshold
-                / (self.tolerance_factor * self.core_weight_threshold - 1)
-            )
+        self._time_period = math.ceil(
+            (1 / self.decaying_factor)
+            * math.log((self.mu * self.beta) / (self.mu * self.beta - 1))
         )
+        self._init_buffer = deque()
+        self._n_samples_seen = 0
+
+    @property
+    def centers(self):
+        if self.clusters:
+            return {
+                k: cluster.calc_center(self.timestamp)
+                for k, cluster in self.clusters.items()
+            }
+        else:
+            return {}
```

Comment:
No need for the `else`

Reply:
In fact, if `self.clusters` is an empty `dict`, you don't need the `if` condition

Reply:
+1 to that

Reply:
Nice catch. Much better now.

Thanks for the review!